dora/core/common/src/main/java/alluxio/membership/ServiceDiscoveryRecipe.java

```diff
@@ -0,0 +1,323 @@
+import alluxio.exception.status.AlreadyExistsException;
+import alluxio.resource.LockResource;
+import alluxio.util.ThreadFactoryUtils;
+
+import com.google.common.base.Preconditions;
+import io.etcd.jetcd.ByteSequence;
+import io.etcd.jetcd.KeyValue;
+import io.etcd.jetcd.Txn;
+import io.etcd.jetcd.kv.TxnResponse;
+import io.etcd.jetcd.lease.LeaseKeepAliveResponse;
+import io.etcd.jetcd.op.Cmp;
+import io.etcd.jetcd.op.CmpTarget;
+import io.etcd.jetcd.op.Op;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.PutOption;
+import io.etcd.jetcd.support.CloseableClient;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * ServiceDiscoveryRecipe for etcd, to track health status
+ * of all registered services.
+ */
+public class ServiceDiscoveryRecipe {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final String BASE_PATH = "/ServiceDiscovery";
+  AlluxioEtcdClient mAlluxioEtcdClient;
+  ScheduledExecutorService mExecutor;
+  String mClusterIdentifier = "";
+  // Will look like /ServiceDiscovery/<mClusterIdentifier>
+  String mRegisterPathPrefix = "";
+  @SuppressFBWarnings({"URF_UNREAD_FIELD"})
+  private final ReentrantLock mRegisterLock = new ReentrantLock();
+  final ConcurrentHashMap<String, ServiceEntity> mRegisteredServices = new ConcurrentHashMap<>();
+
+  /**
+   * CTOR for ServiceDiscoveryRecipe.
+   * @param client
+   * @param clusterIdentifier
+   */
+  public ServiceDiscoveryRecipe(AlluxioEtcdClient client, String clusterIdentifier) {
+    mAlluxioEtcdClient = client;
+    mAlluxioEtcdClient.connect();
+    mClusterIdentifier = clusterIdentifier;
+    mRegisterPathPrefix = String.format("%s%s%s", BASE_PATH,
+        MembershipManager.PATH_SEPARATOR, mClusterIdentifier);
+    mExecutor = Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("service-discovery-checker", false));
+    mExecutor.scheduleWithFixedDelay(this::checkAllForReconnect,
+        AlluxioEtcdClient.DEFAULT_LEASE_TTL_IN_SEC, AlluxioEtcdClient.DEFAULT_LEASE_TTL_IN_SEC,
+        TimeUnit.SECONDS);
+  }
+
+  /**
+   * Apply for a new lease for given ServiceEntity.
```

Comment:
add a line in the comment about the thread safety guarantee

Reply:
done